test/puppetlabs/http/client/async_plaintext_test.clj

```diff
@@ -352,4 +356,95 @@
+                                (promise))]
+      (testwebserver/with-test-webserver-and-config
+       (build-content-handler data initial-bytes-read? nil) port {:shutdown-timeout-seconds 1}
+       (with-open [client (async/create-client {:connect-timeout-milliseconds 1000 :socket-timeout-milliseconds 2000})]
+         (let [response @(common/get
+                          client
+                          (str "http://localhost:" port "/hello")
+                          opts)
+               {:keys [status body]} response]
+           (is (= 200 status))
+           (let [instream body
+                 buf (make-array Byte/TYPE 4)
+                 _ (.read instream buf)]
+             ;; Make sure we can read a few chars off of the stream
+             (is (= "xxxx" (String. buf "UTF-8")))
+             ;; Indicate we read some chars
+             (if initial-bytes-read? (deliver initial-bytes-read? true))
+             ;; Read the rest and validate the content
+             (let [final-string (str "xxxx" (slurp instream))]
+               (is (= (str data "xxxx") final-string))))))))))
+
+(defn- clojure-streaming-failure
+  [data-size opts]
+  (testing (str "clojure streaming failure: " data-size " bytes with opts " opts)
+    (let [data (apply str (repeat (- data-size 4) "x"))]
+      (try
+        (testwebserver/with-test-webserver-and-config
+         (build-content-handler data nil true) port {:shutdown-timeout-seconds 1}
+         (with-open [client (async/create-client {:connect-timeout-milliseconds 1000 :socket-timeout-milliseconds 2000})]
+           (let [response @(common/get
+                            client
+                            (str "http://localhost:" port "/hello")
+                            opts)
+                 {:keys [body error]} response]
+             (if (= :unbuffered-stream (:as opts))
+               (do
+                 (if error
+                   ;; If there's an error this should behave the same as :stream
+                   (is (instance? SocketTimeoutException error))
+                   ;; If there is no error, let's consume the body to get the exception
+                   (is (thrown? SocketTimeoutException (slurp body)))))
+               (do
+                 (is error)
+                 (is (instance? SocketTimeoutException error)))))))
+        (catch TimeoutException e
+          ;; Expected whenever a server-side failure is generated
+          )))))
+
+(deftest clojure-streaming
+  (testing "clojure streaming is consistent with different payload sizes and opts"
+    (testlogging/with-test-logging
+     (doseq [data-size [32 (* 64 1024) (* 1024 1024)]
+             decompress-body? [false true]
+             as [:unbuffered-stream :stream]]
+       (clojure-streaming-success data-size {:as as :decompress-body decompress-body?})
+       (clojure-streaming-failure data-size {:as as :decompress-body decompress-body?})))))
```

Comment:
It would be nice to have some variation of the `clojure-streaming-failure` test where we can better guarantee that the client is seeing the `server-side-failure` exception, when thrown, as translating directly to a failure on the client.  With the timing-specific nature of these tests, I think that will happen in some cases but doesn't seem to be guaranteed.  For that, maybe we could use a much longer `socket-timeout-milliseconds`, confirm that we can get the first few bytes off of the stream (like in the `clojure-streaming-success` case), and confirm that we get an appropriate exception from the next client read attempt - lining up with when the `server-side-failure` exception is thrown.

Does this make sense?